sequencer/src/persistence/fs.rs

```diff
@@ -53,6 +64,113 @@ impl Persistence {
     fn anchor_leaf_path(&self) -> PathBuf {
         self.0.join("anchor_leaf")
     }
+
+    fn vid_dir_path(&self) -> PathBuf {
+        self.0.join("vid")
+    }
+
+    fn da_dir_path(&self) -> PathBuf {
+        self.0.join("da")
+    }
+
+    fn high_qc_path(&self) -> PathBuf {
+        self.0.join("high_qc")
+    }
+}
+
+#[async_trait]
+impl Storage<SeqTypes> for Persistence {
+    async fn append_vid(
+        &self,
+        proposal: &Proposal<SeqTypes, VidDisperseShare<SeqTypes>>,
+    ) -> anyhow::Result<()> {
+        let view_number = proposal.data.get_view_number().get_u64();
+        let dir_path = self.vid_dir_path();
+
+        fs::create_dir_all(dir_path.clone()).context("failed to create vid dir")?;
+
+        let file_path = dir_path.join(view_number.to_string()).with_extension("txt");
+
+        let mut file = fs::OpenOptions::new()
+            .append(true)
+            .create(true)
+            .open(file_path)?;
+
+        let proposal_bytes = bincode::serialize(&proposal).context("serialize proposal")?;
+
+        file.write_all(&proposal_bytes)?;
+
+        Ok(())
+    }
+    async fn append_da(
+        &self,
+        proposal: &Proposal<SeqTypes, DAProposal<SeqTypes>>,
+    ) -> anyhow::Result<()> {
+        let view_number = proposal.data.get_view_number().get_u64();
+        let dir_path = self.da_dir_path();
+
+        fs::create_dir_all(dir_path.clone()).context("failed to create da dir")?;
+
+        let file_path = dir_path.join(view_number.to_string()).with_extension("txt");
+
+        let mut file = fs::OpenOptions::new()
+            .append(true)
+            .create(true)
+            .open(file_path)?;
+
+        let proposal_bytes = bincode::serialize(&proposal).context("serialize proposal")?;
+
+        file.write_all(&proposal_bytes)?;
+
+        Ok(())
+    }
+    async fn record_action(&self, view: ViewNumber, action: HotShotAction) -> anyhow::Result<()> {
```

Comment:
I don't love this taking `&self`. The update is non-atomic (first check that the view is newer than the old one, then overwrite the file), so we were relying on the fact that there was only one task in the node writing to this file at a time, which in turn was enforced by `&mut self`. Without that, we may need a `Mutex` protecting this functionality somehow.

Reply:
I'm beginning to wonder if we should implement the HotShot `Storage` trait on `RwLock<SequencerPersistence>` rather than directly on `SequencerPersistence`. This would fix the issue of `SequencerPersistence` requiring `Clone` and the awkwardness of setting `Storage = Arc<RwLock<Storage>>` in the testable data source